From fa98c4010d0c1ffb5348897a22f76a026f283da9 Mon Sep 17 00:00:00 2001
From: Richard Yu <richard.yu@mitrastar.com.tw>
Date: Fri, 2 May 2025 15:46:02 +0800
Subject: [PATCH 6/7] ieee1905_transport: main: Implement TERM Signal Handler
 in ieee1905_transport
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add a signal handler for SIGTERM in the ieee1905_transport process
to support graceful termination. Previously, the process did not
handle TERM signals, resulting in a forced kill (SIGKILL) after a
10-second timeout during prplMesh reloads via the killall_program()
function in prplmesh_utils.sh.

With this update, ieee1905_transport responds to TERM promptly,
expediting the prplMesh reload process and reducing unnecessary
delays.

Additionally, the main() function now consistently returns 0 when the
event loop exits, aligning the ieee1905_transport process’ exit
behaviour with other prplMesh processes such as beerocks_agent_main
and beerocks_master_main.

https://prplfoundationcloud.atlassian.net/browse/PPM-3294

Signed-off-by: Richard Yu <richard.yu@mitrastar.com.tw>
(cherry picked from commit 2f0fe856aa03d1a3a7ed2e97c5671a01fddf352d)
---
 .../transport/ieee1905_transport/main.cpp     | 58 +++++++++++++++++--
 1 file changed, 54 insertions(+), 4 deletions(-)
 mode change 100644 => 100755 framework/transport/ieee1905_transport/main.cpp

diff --git a/framework/transport/ieee1905_transport/main.cpp b/framework/transport/ieee1905_transport/main.cpp
old mode 100644
new mode 100755
index 328748571..ba0d0763a
--- a/framework/transport/ieee1905_transport/main.cpp
+++ b/framework/transport/ieee1905_transport/main.cpp
@@ -32,6 +32,49 @@ using namespace beerocks;
 using namespace beerocks::net;
 using namespace beerocks::transport;
 
+static bool g_running = true;
+static int s_signal   = 0;
+
+static void handle_signal()
+{
+    if (!s_signal)
+        return;
+
+    switch (s_signal) {
+
+    // Terminate
+    case SIGTERM:
+    case SIGINT:
+        LOG(INFO) << "Caught signal '" << strsignal(s_signal) << "' Exiting...";
+        g_running = false;
+        break;
+
+    default:
+        LOG(WARNING) << "Unhandled Signal: '" << strsignal(s_signal) << "' Ignoring...";
+        break;
+    }
+
+    s_signal = 0;
+}
+
+static void init_signals()
+{
+    // Signal handler function
+    auto signal_handler = [](int signum) { s_signal = signum; };
+
+    struct sigaction sigterm_action;
+    sigterm_action.sa_handler = signal_handler;
+    sigemptyset(&sigterm_action.sa_mask);
+    sigterm_action.sa_flags = 0;
+    sigaction(SIGTERM, &sigterm_action, NULL);
+
+    struct sigaction sigint_action;
+    sigint_action.sa_handler = signal_handler;
+    sigemptyset(&sigint_action.sa_mask);
+    sigint_action.sa_flags = 0;
+    sigaction(SIGINT, &sigint_action, NULL);
+}
+
 static std::shared_ptr<EventLoop> create_event_loop()
 {
     // Create application event loop to wait for blocking I/O operations.
@@ -119,6 +162,8 @@ int main(int argc, char *argv[])
     breakpad_ExceptionHandler();
 #endif
 
+    init_signals();
+
     mapf::Logger::Instance().LoggerInit("transport");
 
     /**
@@ -159,11 +204,16 @@ int main(int argc, char *argv[])
      * Run the application event loop
      */
     MAPF_INFO("starting main loop...");
-    int exit_code = 0;
-    while (0 == exit_code) {
+    while (g_running) {
+        // Handle signals
+        if (s_signal) {
+            handle_signal();
+            continue;
+        }
+
         if (event_loop->run() < 0) {
             LOG(ERROR) << "Broker event loop failure!";
-            exit_code = -1;
+            break;
         }
     }
     MAPF_INFO("done");
@@ -174,5 +224,5 @@ int main(int argc, char *argv[])
     ieee1905_transport.stop();
     broker->stop();
 
-    return exit_code;
+    return 0;
 }
-- 
2.34.1

